src/main/scala/io/iohk/ethereum/domain/Transaction.scala

```diff
@@ -0,0 +1,92 @@
+package io.iohk.ethereum.domain
+
+import akka.util.ByteString
+
+import org.bouncycastle.util.encoders.Hex
+
+sealed trait Transaction {
+  def nonce: BigInt
+  def gasPrice: BigInt
+  def gasLimit: BigInt
+  def receivingAddress: Option[Address]
+  def value: BigInt
+  def payload: ByteString
+
+  def isContractInit: Boolean = receivingAddress.isEmpty
+}
+
+object Transaction {
+  val Type01: Int = 1
+  val LegacyThresholdLowerBound: Int = 0xc0
+  val LegacyThresholdUpperBound: Int = 0xfe
+}
+
+sealed trait TypedTransaction extends Transaction
+
+object LegacyTransaction {
+
+  val NonceLength = 32
+  val GasLength = 32
+  val ValueLength = 32
+
+  def apply(
+      nonce: BigInt,
+      gasPrice: BigInt,
+      gasLimit: BigInt,
+      receivingAddress: Address,
+      value: BigInt,
+      payload: ByteString
+  ): LegacyTransaction =
+    LegacyTransaction(nonce, gasPrice, gasLimit, Some(receivingAddress), value, payload)
+
+}
+
+case class LegacyTransaction(
+    nonce: BigInt,
+    gasPrice: BigInt,
+    gasLimit: BigInt,
+    receivingAddress: Option[Address],
+    value: BigInt,
+    payload: ByteString
+) extends Transaction {
+  override def toString: String = {
+    val receivingAddressString =
+      receivingAddress.map(addr => Hex.toHexString(addr.toArray)).getOrElse("[Contract creation]")
+
+    s"LegacyTransaction {" +
+      s"nonce: $nonce " +
+      s"gasPrice: $gasPrice " +
+      s"gasLimit: $gasLimit " +
+      s"receivingAddress: $receivingAddressString " +
+      s"value: $value wei " +
+      s"payload: ${if (isContractInit) "ContractInit: " else "TransactionData: "}${Hex.toHexString(payload.toArray[Byte])} " +
+      s"}"
+  }
+}
+
+case class TransactionWithAccessList(
```

Comment:
Could this by any chance simply be a `Transaction`? :-)

Reply:
it would collide with `trait Transaction`. or do you have a good name for the base one?

Reply:
`TransactionBase` maybe?